File: Android/app/src/main/java/com/example/supplementtracker/service/UpdateService.kt
package com.example.supplementtracker.service

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Thông tin phiên bản ứng dụng từ xa.
 */
data class AppUpdateInfo(
    val version: String,
    val updateUrl: String,
    val forceUpdate: Boolean,
    val releaseNotes: String
)

data class UpdateConfig(
    val latestVersion: String,
    val isForceUpdate: Boolean,
    val releaseNotes: String,
    val updateUrl: String
)

/**
 * Dịch vụ kiểm tra cập nhật phiên bản (Android).
 */
class UpdateService(
    context: Context
) {
    private val prefs = OakPrefs.get(context)
    private val _updateInfo = MutableStateFlow<AppUpdateInfo?>(null)
    val updateInfo = _updateInfo.asStateFlow()

    private val _isUpdateAvailable = MutableStateFlow(false)
    val isUpdateAvailable = _isUpdateAvailable.asStateFlow()

    /**
     * Kiểm tra phiên bản mới bất đồng bộ.
     */
    suspend fun checkForUpdates(currentVersionName: String) {
        val config = fetchConfig() ?: return
        if (!isVersionNewer(config.latestVersion, currentVersionName)) return
        if (!config.isForceUpdate && getSkippedUpdateVersion() == config.latestVersion) return

        withContext(Dispatchers.Main) {
            _updateInfo.value = AppUpdateInfo(
                version = config.latestVersion,
                updateUrl = config.updateUrl,
                forceUpdate = config.isForceUpdate,
                releaseNotes = config.releaseNotes
            )
            _isUpdateAvailable.value = true
        }
    }

    fun dismissUpdate() {
        _isUpdateAvailable.value = false
    }

    fun skipUpdate(version: String) {
        if (version.isBlank()) {
            dismissUpdate()
            return
        }
        prefs.edit().putString(KEY_SKIPPED_VERSION, version).apply()
        dismissUpdate()
    }

    private fun getSkippedUpdateVersion(): String? {
        return prefs.getString(KEY_SKIPPED_VERSION, null)
    }

    // ponytail: only trust update URLs from known hosts.
    private val allowedHosts = listOf("github.com", "githubusercontent.com", "oakhealthy.com")

    private fun isUpdateUrlAllowed(urlString: String): Boolean {
        return runCatching {
            val host = URL(urlString).host?.lowercase() ?: return@runCatching false
            allowedHosts.any { host.endsWith(it) }
        }.getOrDefault(false)
    }

    private suspend fun fetchConfig(): UpdateConfig? = withContext(Dispatchers.IO) {
        runCatching {
            val url = URL(CONFIG_URL)
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = 5_000
                readTimeout = 5_000
            }

            connection.inputStream.bufferedReader().use { reader ->
                val json = JSONObject(reader.readText())
                val updateUrl = json.getString("update_url")
                if (!isUpdateUrlAllowed(updateUrl)) return@withContext null
                UpdateConfig(
                    latestVersion = json.getString("latest_version"),
                    isForceUpdate = json.getBoolean("is_force_update"),
                    releaseNotes = json.getString("release_notes"),
                    updateUrl = updateUrl
                )
            }
        }.getOrNull()
    }

    private fun isVersionNewer(remote: String, current: String): Boolean {
        return compareVersions(remote, current) > 0
    }

    private fun compareVersions(a: String, b: String): Int {
        val aParts = a.split(".").map { it.toIntOrNull() ?: 0 }
        val bParts = b.split(".").map { it.toIntOrNull() ?: 0 }
        val count = maxOf(aParts.size, bParts.size)

        for (i in 0 until count) {
            val left = aParts.getOrElse(i) { 0 }
            val right = bParts.getOrElse(i) { 0 }
            if (left != right) return left - right
        }
        return 0
    }

    private companion object {
        private const val PREFS_NAME = "oak_update_prefs"
        private const val KEY_SKIPPED_VERSION = "SkippedUpdateVersion"
        private const val CONFIG_URL =
            "https://gist.githubusercontent.com/QuachGia1994/901e36f6bab91729d5dd0e2ccce7202f/raw/oak_update.json"
    }
}

File: iOS/Services/UpdateService.swift
import Foundation
import Observation

/// Thông tin phiên bản ứng dụng từ xa.
public struct AppUpdateInfo: Codable, Sendable {
    public let version: String
    public let updateUrl: String
    public let forceUpdate: Bool
    public let releaseNotes: String
}

public struct UpdateConfig: Codable, Sendable {
    public let latestVersion: String
    public let isForceUpdate: Bool
    public let releaseNotes: String
    public let updateUrl: String
    
    enum CodingKeys: String, CodingKey {
        case latestVersion = "latest_version"
        case isForceUpdate = "is_force_update"
        case releaseNotes = "release_notes"
        case updateUrl = "update_url"
    }
}

/// Dịch vụ kiểm tra cập nhật phiên bản (iOS).
@Observable
@MainActor
public final class UpdateService {
    public var updateInfo: AppUpdateInfo?
    public var isUpdateAvailable: Bool = false
    
    public init() {}
    
    private static let allowedUpdateHosts = ["apps.apple.com", "testflight.apple.com"]

    /// Kiểm tra phiên bản mới từ GitHub Gist/Remote JSON.
    public func checkForUpdates() async {
        guard let url = URL(string: "https://gist.githubusercontent.com/QuachGia1994/901e36f6bab91729d5dd0e2ccce7202f/raw/oak_update.json") else { return }
        let currentVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
        
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let config = try JSONDecoder().decode(UpdateConfig.self, from: data)
            guard isVersion(config.latestVersion, newerThan: currentVersion) else { return }
            if config.isForceUpdate == false, skippedUpdateVersion == config.latestVersion { return }
            guard isTrustedUpdateURL(config.updateUrl) else { return }
            
            updateInfo = AppUpdateInfo(
                version: config.latestVersion,
                updateUrl: config.updateUrl,
                forceUpdate: config.isForceUpdate,
                releaseNotes: config.releaseNotes
            )
            isUpdateAvailable = true
        } catch {
            return
        }
    }

    private func isTrustedUpdateURL(_ urlString: String) -> Bool {
        guard let url = URL(string: urlString), let host = url.host else { return false }
        return Self.allowedUpdateHosts.contains(host)
    }

    public func skipUpdate(version: String) {
        guard version.isEmpty == false else {
            isUpdateAvailable = false
            return
        }
        UserDefaults.standard.set(version, forKey: Self.skippedVersionKey)
        isUpdateAvailable = false
    }

    private var skippedUpdateVersion: String? {
        UserDefaults.standard.string(forKey: Self.skippedVersionKey)
    }
    
    private func isVersion(_ a: String, newerThan b: String) -> Bool {
        compareVersions(a, b) == .orderedDescending
    }
    
    private func compareVersions(_ a: String, _ b: String) -> ComparisonResult {
        let aParts = a.split(separator: ".").map { Int($0) ?? 0 }
        let bParts = b.split(separator: ".").map { Int($0) ?? 0 }
        let count = max(aParts.count, bParts.count)
        
        for i in 0..<count {
            let left = i < aParts.count ? aParts[i] : 0
            let right = i < bParts.count ? bParts[i] : 0
            if left != right { return left > right ? .orderedDescending : .orderedAscending }
        }
        return .orderedSame
    }

    private static let skippedVersionKey = "SkippedUpdateVersion"
}
